### DISCUSSION TITLE
Return m expression text for all queries in semantic model

### DISCUSSION THREAD

**Discussion by ebeery11** [microsoft/semantic-link-labs#307] [Q&A] [answered]
I feel like this must already be possible with semantic link, but I am struggling to find any mention in the documentation of a way to return the m expression text for all queries within a semantic model.

The closest I've gotten so far is using sempy `fabric.get_tmsl` but parsing this from the entire TMSL seems like overkill. 

I'm also not seeing anything in the `sempy_labs.tom` that seems to do exactly this.

**Comment by ebeery11** [comment]
Whoops, I think I answered my own question.  

Looks like the 'Query' column of `fabric.list_partitions` was what I was looking for.   

I'm guessing `sempy_labs.tom` `all_partitions()` will return something similar.

**Comment by m-kovalsky** [comment, marked as answer]
This is the easiest way to access that.
```python
import sempy.fabric as fabric
fabric.list_partitions(dataset='', workspace='')
```

**Reply by ebeery11** [reply]
Thanks @m-kovalsky.  One more question to clarify that I'm getting myself confused on:  

Do basically all of these sempy functions require that the dataset reside in a Premium workspace?  I'm getting a "user does not have permission to call the Discover method" error when attempting to use functions like `fabric.list_datasets()` or `fabric.list_partitions()` on non-Premium workspaces, but it's a bit confusing to me because `fabric.list_workspaces()` returns _all_ workspaces, including those not on dedicated capacities.

**Comment by m-kovalsky** [comment]
Premium is not a requirement. What is your permission on the workspace? You must be be either admin/member/contributor.

**Reply by ebeery11** [reply]
Hmm.  I am admin on all of the ones that I've been trying.  Maybe I am missing something else...

**Comment by m-kovalsky** [comment]
Can you check if this tenant setting is disabled?

![image](https://github.com/user-attachments/assets/3037e4fe-3b99-4f51-9652-93625b106f69)

**Reply by ebeery11** [reply]
I'm not a tenant admin, but checking with them now to see. Will report back. Thanks!

**Reply by ebeery11** [reply]
@m-kovalsky I've confirmed that this setting is disabled in our tenant.

**Comment by m-kovalsky** [comment]
Do you have xmla r/w enabled on the capacity?

You can probably run list_datasets if you set the parameter mode=‘rest’ because it doesn’t use xmla (that forces the function to call an api, not xmla).

**Reply by ebeery11** [reply]
Thanks.  If by "the capacity" you mean the Premium capacity, I think the answer is yes, because I'm able to do things like trigger refreshes, etc. using XMLA through Tabular Editor (on models residing in Premium). 

I confirmed that I could run `list_datasets` with `mode='rest'` on a model in a Pro workspace.  But I'm assuming `list_partitions` doesn't work the same way because there's no REST equivalent?

**Comment by m-kovalsky** [comment]
Correct. The list functions for the semantic model (list_tables, list_columns, list_partitions etc) use the Tabular Object Model API (TOM) to get the metadata. The odd thing is that I’m not aware of any calls to ‘discover’ objects.

**Reply by ebeery11** [reply]
Thanks for all the info.  Ok, I think I'll have to give up for now.  Was hoping to use this to do a quick audit of our team's m code to trace impact analysis of some upstream database changes.  Will have to just scream test instead 🤷‍♂️

In case it's of any future interest, here's the last call in the error message that I'm getting when attempting to use `list_partitions` on a non-Premium model.  Appears to be failing on the `tom_server.Connect()` function.

```
File ~/cluster-env/trident_env/lib/python3.10/site-packages/sempy/fabric/_client/_utils.py:57, in _create_tom_server(connection_string, token_provider)
     54 tom_server.AccessToken = get_access_token(None)
     55 tom_server.OnAccessTokenExpired = Func[AccessToken, AccessToken](get_access_token)
---> 57 tom_server.Connect(connection_string)
     59 return tom_server

OperationException: The '<euii>MYUSERNAME</euii>' user does not have permission to call the Discover method.
```

**Reply by ebeery11** [reply]
@m-kovalsky just following up on this as I had another use case.  I hacked together a solution(?) using your `get_semantic_model_definition` function, decoding the base64 file output into json, and unpacking the json to return just the query expressions.   

Probably there is a better way, but it seems to work(ish) for now on both Premium and Pro models.  Unfortunately, I frequently hit the 8000-character limit for storing in a Lakehouse table.

**Reply by m-kovalsky** [reply]
You’re much better off finding a way to use fabric.list_partitions() as otherwise you will have to reinvent the wheel several times over.

**Reply by ebeery11** [reply]
@m-kovalsky definitely agree, but the fact that `fabric.list_partitions()` doesn't work for non-Premium models seems to be a hard limitation for now

**Reply by m-kovalsky** [reply]
Are you storing a single M expression in a given row of the lakehouse table? 8000 characters seems plenty if that’s the case so sounds like the M code is overdone.

**Reply by ebeery11** [reply]
Yes I am.  Agree, 8000 is plenty for most cases, but not all.  In my initial screening, about ~1% of our queries (11/1246) had >8000 characters and so were truncated.  

In this case even with truncation, it's still helpful, since what I'm mostly after is tracing SQL table/view dependency, and those are typically in the first few lines of the m expression.
